secure/storage/vault/src/lib.rs

```diff
@@ -136,36 +194,185 @@ impl Client {
+/// Below is a sample output of ReadSecretListResponse. All fields are decoded and used.
+/// Note: in the case that a secret contains a subpath, that will be returned. Vault does
+/// not automatically recurse.
+/// {
+///   "data": {
+///     "keys": ["foo", "foo/"]
+///   }
+/// }
 #[derive(Debug, Deserialize, PartialEq, Serialize)]
 struct ReadSecretListResponse {
     data: ReadSecretListData,
 }
 
+/// See ReadSecretListResponse
 #[derive(Debug, Deserialize, PartialEq, Serialize)]
 struct ReadSecretListData {
     keys: Vec<String>,
 }
 
+/// Below is a sample output of ReadSecretResponse. Note, this returns all keys within a secret.
+/// Only fields leveraged by this framework are decoded.
+/// {
+///   "data": {
+///     "data": {
+///       "foo": "bar"
+///     },
+///     "metadata": {
+///       "created_time": "2018-03-22T02:24:06.945319214Z",
+///       "deletion_time": "",
+///       "destroyed": false,
+///       "version": 1
+///     }
+///   }
+/// }
 #[derive(Debug, Deserialize, PartialEq, Serialize)]
 struct ReadSecretResponse {
     data: ReadSecretData,
 }
 
+/// See ReadPolicyResponse
 #[derive(Debug, Deserialize, PartialEq, Serialize)]
 struct ReadSecretData {
     data: HashMap<String, String>,
 }
 
+/// This data structure is used to represent both policies read from Vault and written to Vault.
+/// Thus the same Policy read, can then be written back after amending. Vault stores the rules or
+/// per path policies in an encoded json blob, so that effectively means json within json, hence
+/// the unusual semantics below.
+/// {
+///   rules: json!{
+///     path: {
+///       'auth/*': { capabilities: ['create', 'read', 'update', 'delete', 'list', 'sudo'] },
+///       'sys/auth/*': { capabilities: ['create', 'read', 'update', 'delete', 'sudo'] },
+///     }
+///   }
+/// }
+/// Note: Vault claims rules is deprecated and policy should be used instead, but that doesn't seem
+/// to work and makes the reading asymmetrical from the writing.
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct Policy {
+    #[serde(skip)]
+    internal_rules: PolicyPaths,
+    rules: String,
+}
+
+impl Policy {
+    pub fn new() -> Self {
+        Self {
+            internal_rules: PolicyPaths {
+                path: HashMap::new(),
+            },
+            rules: "".to_string(),
+        }
+    }
+
+    pub fn add_policy(&mut self, path: &str, capabilities: Vec<Capability>) {
+        let path_policy = PathPolicy { capabilities };
+        self.internal_rules
+            .path
+            .insert(path.to_string(), path_policy);
+    }
+}
+
+impl TryFrom<serde_json::Value> for Policy {
+    type Error = serde_json::Error;
+
+    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
+        let mut policy: Self = serde_json::from_value(value)?;
+        policy.internal_rules = serde_json::from_str(&policy.rules)?;
+        Ok(policy)
+    }
+}
+
+impl TryInto<serde_json::Value> for Policy {
```

Comment:
the idiomatic way is `impl TryFrom<Policy> for serde_json::Value` and if you don't want clone use `impl TryFrom<&Policy> for serde_json::Value`